WordPress/gutenberg
# BlockSettingsMenu: Ensure only one block settings menu is open at a time

## What?


Fixes https://github.com/WordPress/gutenberg/issues/50988 and unblocks #50273.

Ensure only one BlockSettingsMenu is open at a time by adding state with the `clientId` of the block settings menu most recently opened. All other open block settings menus will close if they have a `clientId` and it does not match the id of the menu currently open.

## Why?


This resolves an issue where multiple block settings menus can be open at one time if the current window does not have focus. While this sounds like an edge case, it's fairly common on Macs when users right click into a window that is not currently active, which is an issue that's revealed when we go to attempt to add in right-click support as in #50273. However, it's also an (albeit very subtle) issue for cases where a user CMD+clicks from an active window into an inactive one. This will fire a click into a window without ever shifting focus there, so the current `onFocusOutside` logic isn't quite sufficient for the block settings menu.

## How?


If we introduce some state in the block editor store to keep track of the latest `clientId` of the most recently opened block settings menu, then we can auto-close any block settings menus that are already opened.

* Add state to track the id of the most recently opened block settings menu
* Allow `onToggle` and `open` to be passed down so that we can:
  * Add a callback so that we can perform an action when the dropdown menu is toggled open (so we can then update state).
  * Pass down an explicit `open` state so that we can treat the dropdowns as controlled components.

This PR updates the block editor package, the necessary updates to the `Dropdown` and `DropdownMenu` components were made in https://github.com/WordPress/gutenberg/pull/54257.

## Testing Instructions





1. On a Mac, with focus in a window that is not your test site, hover the mouse cursor over an inactive window that's open to the post or site editors and that also has the list view open.
2. CMD+Click on the area of list view items where the ellipsis block settings menu button should be.
3. Move the mouse cursor up to another list view item and CMD+Click on where the ellipsis menu button should be.
4. Prior to this PR, multiple instances of the dropdown menu will be open.
5. With this PR applied, any already open menus should automatically close.

Note: while the issue in testing here is quite subtle, it's a good idea for us to try to fix it prior to introducing right click in #50273 as it's fairly common for folks to click between two browser windows — and (on a Mac at least) right click does not change focus between windows.

### Testing Instructions for Keyboard


## Screenshots or screencast 

This is how it can look if a user has multiple block settings menus open (prior to this PR):

![image](https://github.com/WordPress/gutenberg/assets/14988353/bd858c13-5916-4468-b68a-4bfbe6445254)

| Before | After |
| --- | --- |
| ![2023-08-31 16 27 43](https://github.com/WordPress/gutenberg/assets/14988353/2364d4ac-9326-491d-942a-7c63a2ebe893) | ![2023-08-31 16 25 34](https://github.com/WordPress/gutenberg/assets/14988353/5e500a28-a411-4030-bf24-960615c646eb) |
